pkg/controllers/management/clusterdeploy/clusterdeploy.go

```diff
@@ -110,6 +132,24 @@ func (cd *clusterDeploy) doSync(cluster *v3.Cluster) error {
 }
 
 func redeployAgent(cluster *v3.Cluster, desiredAgent, desiredAuth string) bool {
+	if cluster.Status.AgentImageOverride != "" {
+		// overridden agent
+		if cluster.Status.AgentImageOverride == cluster.Spec.DesiredAgentImage {
+			return false
+		}
+		if cluster.Status.AgentImageOverride == cluster.Status.AgentImage {
+			return false
+		}
+
+		// kick a redeploy with the overriden agent
+		logrus.Infof("Redeploy Rancher Agents due to Agent Image Override for %s: was %s and will be %s",
+			cluster.Name, cluster.Status.AgentImage, cluster.Status.AgentImageOverride)
+		clearAgentImages(cluster.Name)
+		cluster.Status.AgentImage = ""
+		cluster.Spec.DesiredAgentImage = cluster.Status.AgentImageOverride
```

Comment:
This function used to be a side-effect free evaluator of whether or not you should redeploy an agent. It now has the side effect of setting two fields on the cluster object. I think this is a bad idea and I think the function should remain side-effect free. Its also not clear to me that setting this field is going to have the effect that you want. Will discuss in person.